### PR TITLE
[Backport 1.21] TLS: Allow specification of both typed and non-typed san matchers in config (20529)

### PR DESCRIPTION
**Pull request by pradeepcrao**
Signed-off-by: Pradeep Rao <pcrao@google.com>

Commit Message:
Additional Description:
Backport #20529 to 1.21
Risk Level: Low
Testing: Modified test.
Docs Changes:
Release Notes:
Platform Specific Features:

### CODE REVIEW

**Review by ggreenway — APPROVED** [review]

**Comment by ggreenway** [comment]
CI is failing
/wait

**Comment by pradeepcrao** [comment]
> CI is failing /wait

Can't reproduce the build failure locally. Doing a merge to kick off tests again.

File: docs/root/version_history/current.rst [hunk 1]
```diff
@@ -9,6 +9,10 @@ Minor Behavior Changes
 ----------------------
 *Changes that may cause incompatibilities for some users, but should not for most*
 
+* cryptomb: remove RSA PKCS1 v1.5 padding support.
+* perf: ssl contexts are now tracked without scan based garbage collection and greatly improved the performance on secret update.
```
Comment (adisuissa) [inline docs/root/version_history/current.rst:13]: These changes don't seem related to the current backport PR.

**Review by adisuissa — CHANGES_REQUESTED** [review]
Overall LGTM. Left a small nit
/lgtm api
/wait

File: docs/root/version_history/current.rst (hunk 1 above)
Comment (pradeepcrao) [reply docs/root/version_history/current.rst:13]: Aah, this happened because of this: https://github.com/envoyproxy/envoy/pull/21226/files

Fixing.

**Review by adisuissa — APPROVED** [review]
LGTM, thanks!
